fgpyo/sam/__init__.py

```diff
@@ -1036,55 +1043,119 @@ class ReadEditInfo:
     """
     Constructs a `ReadEditInfo` instance giving summary stats about how the read aligns to the
-    reference.  Computes the number of mismatches, indels, indel bases and the SAM NM tag.
-    The read must be aligned.
+    reference.  Computes the number of mismatches, indels, indel bases as well as the
+    SAM NM and MD tags.
+
+    Calculation of NM and MD tags is based off of htsjdk:
+    https://github.com/samtools/htsjdk/blob/7034b33636b4cb9fec300a2136588e7c12c7ccd5/src/main/java/htsjdk/samtools/util/SequenceUtil.java#L964:L1029
+
+    Per the SAM specification (https://samtools.github.io/hts-specs/SAMtags.pdf), the NM tag
+    encapsulates the number of differences between the query read and reference sequence, counting
+    only A, C, G and T bases (case-insensitive). Everything else should be considered a mismatch
+    (e.g., ambiguity codes like R and N). We set the default of `n_as_match` to False to be
+    concordant with the SAM specification. Conversely, `htsjdk` treats an N->N as a match.
+
+    If the read is unmapped or the query sequence contains missing bases (`*`), returns None, as it
+    is not possible to recalculate the MD and NM tags without access to the query sequence and
+    reference sequence.
+
+    The order of the CIGAR operator checks is for performance and modeled after htsjdk's
+    `calculateMdAndNmTags`.
 
     Args:
         rec: the read/record for which to calculate values
-        reference_sequence: the reference sequence (or fragment thereof) that the read is
-            aligned to
+        reference_sequence: the reference sequence (or fragment thereof) to which the read is
+            aligned
+        match_htsjdk: if True, mirror htsjdk `calculateMdAndNmTags` -- only match is the bases are
+            equal, including ambiguity codes (e.g., R->R is counted as a match, but R->A is not a
+            match). If False, follow SAM spec (everything else should be considered a mismatch,
+            including ambiguity codes like R and N).
         reference_offset: if provided, assume that reference_sequence[reference_offset] is the
             first base aligned to in reference_sequence, otherwise use r.reference_start
 
     Returns:
         a ReadEditInfo with information about how the read differs from the reference
     """
-    assert not rec.is_unmapped, f"Cannot calculate edit info for unmapped read: {rec}"
+    if rec.is_unmapped or rec.query_sequence is None or rec.query_sequence == NO_QUERY_BASES:
+        return None
 
-    query_offset = 0
-    target_offset = reference_offset if reference_offset is not None else rec.reference_start
-    cigar = Cigar.from_cigartuples(rec.cigartuples)
+    query_offset: int = 0
+    target_offset: int = reference_offset if reference_offset is not None else rec.reference_start
+    cigar: Cigar = Cigar.from_cigartuples(rec.cigartuples)
 
     matches, mms, insertions, ins_bases, deletions, del_bases = 0, 0, 0, 0, 0, 0
```

Comment:
chore: while we're here, can we update this variable to a longer form?
```suggestion
    matches, mismatches, insertions, ins_bases, deletions, del_bases = 0, 0, 0, 0, 0, 0
```